internal/settings/gandi.go

```diff
@@ -0,0 +1,199 @@
+}
+
+func (g *gandi) DNSLookup() bool {
+	return g.dnsLookup
+}
+
+func (g *gandi) IPVersion() models.IPVersion {
+	return g.ipVersion
+}
+
+func (g *gandi) BuildDomainName() string {
+	return buildDomainName(g.host, g.domain)
+}
+
+func (g *gandi) HTML() models.HTMLRow {
+	return models.HTMLRow{
+		Domain:    models.HTML(fmt.Sprintf("<a href=\"http://%s\">%s</a>", g.BuildDomainName(), g.BuildDomainName())),
+		Host:      models.HTML(g.Host()),
+		Provider:  "<a href=\"https://www.gandi.net/\">gandi</a>",
+		IPVersion: models.HTML(g.ipVersion),
+	}
+}
+
+func (g *gandi) setHeaders(request *http.Request) {
+	setUserAgent(request)
+	setContentType(request, "application/json")
+	setAccept(request, "application/json")
+	request.Header.Set("X-Api-Key", g.key)
+}
+
+func (g *gandi) getRecordIP(ctx context.Context, recordType string, client *http.Client) (
+	recordIP string, err error) {
+	u := url.URL{
+		Scheme: "https",
+		Host:   "dns.api.gandi.net",
+		Path:   fmt.Sprintf("/api/v5/domains/%s/records/%s/%s", g.domain, g.host, recordType),
+	}
+
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
+	if err != nil {
+		return "", err
+	}
+	g.setHeaders(request)
+
+	response, err := client.Do(request)
+	if err != nil {
+		return "", err
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("%w: %d: %s",
+			ErrBadHTTPStatus, response.StatusCode, bodyToSingleLine(response.Body))
+	}
+
+	decoder := json.NewDecoder(response.Body)
+	var result struct {
+		Type   string   `json:"rrset_type"`
+		TTL    int      `json:"rrset_ttl"`
+		Name   string   `json:"rrset_name"`
+		Href   string   `json:"rrset_href"`
+		Values []string `json:"rrset_values"`
+	}
+	if err = decoder.Decode(&result); err != nil {
+		return "", fmt.Errorf("%w: %s", ErrUnmarshalResponse, err)
+	}
+	if len(result.Values) == 0 {
+		return "", ErrNoResultReceived
+	}
+	return result.Values[0], nil
+}
+
+func (g *gandi) Update(ctx context.Context, client *http.Client, ip net.IP) (newIP net.IP, err error) {
+	recordType := A
+	if ip.To4() == nil { // IPv6
+		recordType = AAAA
+	}
+
+	recordIP, err := g.getRecordIP(ctx, recordType, client)
+	if err != nil && err != ErrNoResultReceived { // if no ip was defined before, let's proceed with the update
+		return nil, fmt.Errorf("%s: %w", ErrGetRecordIP, err)
+	}
+
+	oldIP := net.ParseIP(recordIP)
+	if ip.Equal(oldIP) { // success, nothing to change
+		return ip, nil
+	}
+
+	u := url.URL{
+		Scheme: "https",
+		Host:   "dns.api.gandi.net",
+		Path:   fmt.Sprintf("/api/v5/domains/%s/records/%s/%s", g.domain, g.host, recordType),
+	}
+
+	buffer := bytes.NewBuffer(nil)
+	encoder := json.NewEncoder(buffer)
+	requestData := struct {
+		Values [1]string `json:"rrset_values"`
+		TTL    int       `json:"rrset_ttl"`
+	}{
+		Values: [1]string{ip.To4().String()},
+		TTL: func() int {
+			ttl := DefaultTTL
+			if g.ttl != 0 {
+				ttl = g.ttl
+			}
+			return ttl
+		}(),
```

Comment:
this is fancy but please just put it above the `requestData := ...` as:

```go
const defaultTTL = 3600
ttl := defaultTTL
if g.ttl != 0 {
	ttl = g.ttl
}
```

to keep it simple, thanks!